### PR TITLE
[#3168] - assign datapoints to assignments by name [no api included yet]

### PR DESCRIPTION
**Pull request by marvinkome**
Currently all data are stored in memory and not yet connected to the API.

#### Screenshots (if appropriate)
datapoints

#### Reviewer Checklist
* [ ] Added an explanation about the work done
* [ ] Connected the PR and the issue on Zenhub
* [ ] Added a test plan to the issue
* [ ] Updated the copyright header (when relevant)
* [ ] Formatted the code
* [ ] Added a documentation (if relevant)
* [ ] Added some unit tests (if relevant)

### CODE REVIEW

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  onSelectDatapoint = (id, checked) => {
+    const { selectedDatapoints } = this.state;
+    let newSelectedDatapoints = [];
+
+    if (checked) {
+      newSelectedDatapoints = selectedDatapoints.concat(id);
+    } else {
+      newSelectedDatapoints = selectedDatapoints.filter(
+        device => device !== id
```
Comment (muloem) [inline Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx]: This looks strange to me... does `device` actually be `dataPointId`?

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+  onSearch = e => {
+    e.preventDefault();
+
+    const searchKey = e.target.searchDatapoints.value;
+    this.context.actions.findDatapoints(searchKey);
+  };
+
+  onSelectDatapoint = (id, checked) => {
+    const { selectedDatapoints } = this.state;
```
Comment (muloem) [inline Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx]: maybe its more accurate to call it `selectedDataPointIds` that way we actually know whats in the array

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    this.setState({ selectedDatapoints: newSelectedDatapoints });
+  };
+
+  addToAssignment = () => {
+    const { selectedDatapoints } = this.state;
+    const { datapoints } = this.context.data;
+    const { addDatapointsToAssignment } = this.context.actions;
+
+    const formattedSelectedDps = selectedDatapoints.map(sDp =>
```
Comment (muloem) [inline Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx]: its not clear to me why the _formatted_ here

Comment (muloem) [reply Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx]: So from what I can gather what you are calling `formattedSelectedDps` are the actual data points which is an array of objects whereas the `selectedDatapoints` is an array of numbers that represent the ids of the datapoints.

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/MainBody.jsx [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
 import DevicesSection from './DevicesSection';
 import SidebarDropdown from './__partials/SidebarDropdown';
 
 import AssignmentsContext from './assignment-context';
 
 export default class AssignmentMain extends React.Component {
   state = {
-    currentTab: 'FORMS',
+    currentTab: 'ASSIGN_DATAPOINTS',
+    selectedDevice: '150482013',
```
[the remark this answers is not in the document]
Comment (marvinkome) [reply Dashboard/app/js/lib/components/devices/AssignmentsEditView/MainBody.jsx]: Yes, would remove that now

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx (hunk 3 above)
Comment (marvinkome) [reply Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx]: Yes. But I have to clarify that too

**Comment by valllllll2000** [comment]
The code Looks good to me but I have seen there are lost of Codacy errors https://app.codacy.com/manual/Akvo/akvo-flow/pullRequest?bid=15721359&prid=4706082&utm_campaign=new_commit&utm_medium=Email&utm_source=Internal&
My knowledge of Javascript does not allow me to tell if they are legit or not but @marvinkome I have invited you to that project in case you want to check them out.

**Comment by marvinkome** [comment]
@valllllll2000 they look more like repetitive formatting problems.

**Comment by marvinkome** [comment]
True. Forgot it was implement. I'll update the PR to match that. 👍🏼

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/MainBody.jsx [hunk 5]
```diff
@@ -12,19 +15,18 @@ import AssignmentsContext from './assignment-context';
...
   };
 
   getDeviceGroups = () => {
     // filter out selected devices
     const { devices, selectedDeviceIds } = this.context.data;
 
-    const filteredDevices = devices.filter(device =>
-      selectedDeviceIds.includes(device.id)
-    );
+    const filteredDevices = devices.filter(device => selectedDeviceIds.includes(device.id));
```
[the remark this answers is not in the document]
Comment (marvinkome) [reply Dashboard/app/js/lib/components/devices/AssignmentsEditView/MainBody.jsx]: Yes

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/index.jsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    const { devices, selectedDatapoints } = this.context.data;
+
+    const deviceData = devices.find(device => device.id === selectedDevice);
+    const selectedDatapoint = selectedDatapoints.find(
+      dp => dp.deviceId === selectedDevice
+    );
+    let datapointsData = [];
+
+    if (selectedDatapoint) {
+      datapointsData = selectedDatapoint.datapoints;
```
[the remark this answers is not in the document]
Comment (marvinkome) [reply Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/index.jsx]: `selectedDatapoint` is same as `datapointAssignment`. Renaming it.

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx [hunk 7]
```diff
@@ -45,13 +45,14 @@ export default class SearchDatapoints extends React.Component {
...
   render() {
+    const { strings } = this.context;
     const { datapoints } = this.context.data;
     const { selectedDatapointsIds } = this.state;
 
     return (
       <div className="search-datapoints">
         <div className="header">
-          <p>Assign datapoints by name</p>
+          <p>{strings.assignDatatpointByNameOrId}</p>
```
Comment (muloem) [inline Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx]:
```suggestion
          <p>{strings.assignDatapointByNameOrId}</p>
```

File: Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx [hunk 8]
```diff
@@ -64,7 +65,11 @@ export default class SearchDatapoints extends React.Component {
           {/* search bar */}
           <form className="search-bar" onSubmit={this.onSearch}>
             <i className="fa fa-search" />
-            <input type="search" id="searchDatapoints" placeholder="Search datapoint by name" />
+            <input
+              type="search"
+              id="searchDatapoints"
+              placeholder={strings.searchDatatpointByNameOrId}
```
Comment (muloem) [inline Dashboard/app/js/lib/components/devices/AssignmentsEditView/screens/AssignDatapoints/SearchDatapoints.jsx]:
```suggestion
              placeholder={strings.searchDatapointByNameOrId}
```

File: Dashboard/app/js/lib/views/devices/assignment-edit-views.jsx [hunk 9]
```diff
@@ -556,5 +568,46 @@ FLOW.AssignmentEditView = FLOW.ReactComponentView.extend(
...
+        return {
+          name: datapoint.get('displayName'),
+          id: datapoint.get('id'),
+        };
+      });
+
+      this.renderReactSide();
+    },
+
+    addDatapointsToAssignment(datapoints, deviceId) {
```
Comment (muloem) [inline Dashboard/app/js/lib/views/devices/assignment-edit-views.jsx]:
```suggestion
    assignDataPointsToDevice(datapoints, deviceId) {
```

File: Dashboard/app/js/lib/views/devices/assignment-edit-views.jsx [hunk 10]
```diff
@@ -556,5 +568,46 @@ FLOW.AssignmentEditView = FLOW.ReactComponentView.extend(
...
+      const selectedDps = this.datapointAssignments;
```
Comment (muloem) [inline Dashboard/app/js/lib/views/devices/assignment-edit-views.jsx]:
```suggestion
      const datapointAssignments = this.datapointAssignments;
```

File: Dashboard/app/js/lib/views/devices/assignment-edit-views.jsx [hunk 11]
```diff
@@ -556,5 +568,46 @@ FLOW.AssignmentEditView = FLOW.ReactComponentView.extend(
...
+      const selectedDp = selectedDps.find(sDp => sDp.deviceId === deviceId);
```
Comment (muloem) [inline Dashboard/app/js/lib/views/devices/assignment-edit-views.jsx]:
```suggestion
      const dataPointAssignment = selectedDps.find(sDp => sDp.deviceId === deviceId);
```

**Pull request merged**